doc/evp_skey.md

```diff
@@ -0,0 +1,90 @@
+EVP_SKEY Design
+===============
+
+We currently presume that symmetric keys are represnted as octet buffer. It may
+be not relevant for the situation of hardware-backed providers, such as
+PKCS#11 or TPM.
+
+This is a design of using opaque usage in symmetric encryption.
+
+
+Creation
+--------
+
+```
+EVP_SKEY * EVP_SKEY_new(void)
+```
+creates an empty `EVP_SKEY` object.
+
+API to set parameters
+---------------------
+
+```
+int EVP_SKEY_fromdata(EVP_CIPHER_CTX *ctx, EVP_SKEY **pskey, OSSL_PARAM params[]);
+int EVP_SKEY_assign(EVP_CIPHER_CTX *ctx, EVP_SKEY *pskey, OSSL_PARAM params[]);
+```
+
+API to derive a symmetric key
+-----------------------------
+
+```
+int EVP_PKEY_derive_SKEY(EVP_PKEY_CTX *ctx, EVP_SKEY *pskey);
+```
+similar to `EVP_PKEY_derive`
+
+We already have the function
+
+```
+EVP_CipherInit_ex2(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
+                   const unsigned char *key, const unsigned char *iv,
+                   int enc, const OSSL_PARAM params[]);
+```
+
+and can introduce
+
+```
+EVP_CipherInit_ex3(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
+                   const EVP_SKEY *skey, const unsigned char *iv,
+                   int enc, const OSSL_PARAM params[]);
```

Comment:
Presumably there would also be EVP_EncryptInit/EVP_DecryptInit variants as well.

Reply:
Or we might avoid it altogether - use EVP_CIPHER_CTX_set1_EVP_SKEY plus EVP_CipherInit_ex2() with NULL key.

Reply:
Done.